File: medium/Two-Pointers/remove-element/remove_element.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int removeElement(vector<int>& nums, int val) {
        /*
         * Remove all target values in-place, return new length.
         * Approach: Slow & Fast Two Pointers
         * Time Complexity: O(n)
         * Space Complexity: O(1)
         */
        // Slow pointer stores valid element index
        int slow = 0;
        // Fast pointer iterates the whole array
        for(int fast = 0; fast < nums.size(); fast++){
            // Save element when it is not equal to target
            if(nums[fast] != val){
                nums[slow++] = nums[fast];
            }
        }
        return slow;
    }
};

File: medium/Two-Pointers/remove-element/remove_element.py
from typing import List

class Solution:
    def removeElement(self, nums: List[int], val: int) -> int:
        """
        Remove target value in-place and return new length.
        Approach: Slow & Fast Two Pointers
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        # Slow pointer: record position of valid elements
        slow = 0
        # Fast pointer: traverse every element in array
        for fast in range(len(nums)):
            # Keep element if it is not the target value
            if nums[fast] != val:
                nums[slow] = nums[fast]
                slow += 1
        # Slow equals the count of remaining elements
        return slow
